extensions/amp-video/0.1/video-cache.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {createElementWithAttributes} from '../../../src/dom';
+import {toArray} from '../../../src/types';
+
+/** @private {?boolean} store the result of whether the document can use cache url */
+let docSupportsCacheResult = null;
+
+/**
+ * Add the caching sources to the video if opted in.
+ * @param {!AmpVideo} video
+ * @return {!Promise}
+ */
+export function addCacheSources(video) {
+  if (!docSupportsCache(video.win)) {
+    return Promise.resolve();
+  }
+  const servicePromise = Services.cacheUrlServicePromiseForDoc(video.element);
+  return servicePromise
+    .then((service) =>
+      Promise.all([
+        service.createCacheUrl(selectVideoSource(video.element)),
+        Services.storyRequestServiceForOrNull(video.win),
+      ])
+    )
+    .then((promiseResult) => {
+      const requestUrl = promiseResult[0].replace('/c/', '/mbv/');
+      return promiseResult[1].executeRequest(requestUrl);
+    })
+    .then((response) => applySourcesToVideo(video.element, response['sources']))
+    .catch(() => {
+      // If cache fails, video should still load properly.
+    });
+}
+
+/**
+ * Selects and returns the prioritized video source URL
+ * @param {!Element} videoEl
+ * @return {string}
+ */
+export function selectVideoSource(videoEl) {
+  const possibleSources = toArray(videoEl.querySelectorAll('source[src]'));
+  if (videoEl.hasAttribute('src')) {
+    possibleSources.push(videoEl);
+  }
+  for (let i = 0; i < possibleSources.length; i++) {
+    if (possibleSources[i].getAttribute('type') === 'video/mp4') {
```

Comment:
This would fail for `source type=" video/mp4"` or `source type="video/mp4;"` or `source type="video/mp4; codecs=...."`